### PR TITLE
Support patches to attached files

### PR DESCRIPTION
**Pull request by olsen232**
Includes some cleanup of apply.py - moving things into helper functions etc.
Still TODO - validate that the patch still applies, ie, that the files being updated haven't changed since the patch was created.

## Related links:

https://github.com/koordinates/kart/issues/583

## Checklist:

- [x] Have you reviewed your own change?
- [x] Have you included test(s)?
- [x] Have you updated the [changelog](https://github.com/koordinates/kart/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: kart/structure.py [hunk 1]
```diff
@@ -308,7 +321,6 @@ def check_values_match_schema(self, repo_diff):
                 if new_schema is None:
                     raise InvalidOperation(
                         f"Can't {feature_delta.type} feature {feature_delta.new_key} in deleted dataset {ds_path}",
-                        exit_code=PATCH_DOES_NOT_APPLY,
```
Comment (craigds) [inline kart/structure.py:311]: why this change?

File: kart/structure.py [hunk 2]
```diff
@@ -235,6 +236,14 @@ def create_tree_from_diff(
                 object_builder.insert(path, blob)
 
         for ds_path, ds_diff in repo_diff.items():
+            if ds_path == FILES_KEY:
+                self.apply_file_diff(
```
Comment (craigds) [inline kart/structure.py:240]: this method should maybe be `apply_files_diff` since it's not just _one_ file, it's all the files

File: kart/apply.py [hunk 3]
```diff
@@ -178,6 +186,59 @@ def _build_signature(patch_metadata, person, repo):
     return repo.author_signature(**signature)
 
 
+def parse_file_diff(file_diff_input, allow_minimal_updates=None):
+    def convert_half_delta(half_delta):
+        if half_delta is None:
+            return None
+        elif half_delta.value.startswith("base64:"):
+            return (half_delta.key, b64decode_str(half_delta.value))
+        else:
```
Comment (craigds) [inline kart/apply.py:195]:
```
"+": "base64:\nbase64 is a way of encoding binary data in strings that you can use in JSON, like this patch."
```

to keep the encoding prefix purely unambiguous maybe we need an enforced prefix for text content? maybe just `:` or `text:`. Ugly though. But it'd be annoying to have a list of prefixes you're not allowed in text data.

**Review by craigds — APPROVED** [review]

File: kart/structure.py (hunk 1 above)
Comment (olsen232) [reply kart/structure.py:311]: hmm good question. I was looking at some exit-codes to make sure they made sense - eg PATCH_DOES_NOT_APPLY should be for the patch is valid, but doesn't work on HEAD, but there should be different error codes for "can't parse the patch". But this one seems pretty reasonably how it is. I'll revert

File: kart/apply.py (hunk 3 above)
Comment (olsen232) [reply kart/apply.py:195]: Yeah I did this prefix thing a bit late and it's a bit ugly - again, it's because there's not really any room for anything out-of-band in the JSON diffs. JSON-LINES on the other hand has more room and it has a "binary" attribute that can disambiguate these two cases. I was hoping we'd just switch to something more like JSON-LINES for everything including patches but we're not really ready for that... this will work for now, until and unless someone starts creating patches which modify text files to have "base64:" at the start of them. It will go away if we ever manage to drop the JSON format.

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,7 +12,7 @@ _When adding new entries to the changelog, please include issue/PR numbers where
 - Bugfix: directing geojson diff output to the console didn't work in a multi-dataset repo, even if only one dataset had changed. [#702](https://github.com/koordinates/kart/issues/702)
 - Improve argument parsing for `kart diff` and `kart show`. [#706](https://github.com/koordinates/kart/issues/706)
 - Bugfix: don't allow `kart merge` to fast-forward if the user specifies a merge message. [#705](https://github.com/koordinates/kart/issues/705)
-- Kart now shows diffs to attached files that have been created using `kart commit-files`, if any [#583](https://github.com/koordinates/kart/issues/583)
+- Kart now shows diffs to attached files that have been created using `kart commit-files`, if any, and patches can edit these files [#583](https://github.com/koordinates/kart/issues/583)
```
[the remark this answers is not in the document]
Comment (olsen232) [reply CHANGELOG.md:15]: Done

File: kart/apply.py (hunk 3 above)
Comment (olsen232) [reply kart/apply.py]: Modified the patch format for text files to start with "text:", but left the other diffs as they were.
